### PR TITLE
apimachinery: add ObjectMeta mutation tests to resttest

### PR DESCRIPTION
**Pull request by sttts**
Test whether ObjectMeta mutation in endpoints (= consumers of registry output) mutates storage cache objects.

This is a regression test for https://github.com/kubernetes/kubernetes/pull/65092:

This is done by preserving shallow copies of the object returned from the underlying etcd storage. After the registry test we can mutate top-level fields of ObjectMeta of these shallow copies and see whether the objects' ObjectMeta returned by the registries is changed as well. It shouldn't. Otherwise, the endpoint handlers (which write ObjectMeta top-level fields) will mutate objects owned by the storage cache.

For most objects a shallow copy does also a shallow copy of ObjectMeta. So this is trivial. But we have e.g. Unstructured where is not automatic. This PR add regression test coverage for this and maybe other future cases.

### CODE REVIEW

**Comment by sttts** [comment]
interceptor does not work like this (obviously). Working on it.

**Comment by sttts** [comment]
@deads2k this is ready. ptal.

**Comment by sttts** [comment]
@liggitt friendly ping

File: staging/src/k8s.io/apiserver/pkg/registry/generic/testing/tester.go [hunk 1]
```diff
@@ -193,10 +222,95 @@ func (t *Tester) emitObject(obj runtime.Object, action string) error {
...
+	defer s.lock.RUnlock()
+	return s.interceptee
+}
+
+// Get unmarshals json found at key into objPtr. On a not found error, will either
+// return a zero object of the requested type, or an error, depending on ignoreNotFound.
+// Treats empty responses and nil response nodes exactly like a not found error.
+// The returned contents may be delayed, but it is guaranteed that they will
+// be have at least 'resourceVersion'.
+func (s *shallowCopyStorageInterceptor) Get(ctx context.Context, key string, resourceVersion string, objPtr runtime.Object, ignoreNotFound bool) error {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/testing/tester.go]: sorry to keep stalling on this... I always get to this point, get concerned about parallel implementations of behavior drifting (one in storage, and one here in test), and get stuck trying to think of an alternative that seems maintainable.

**Pull request closed without merging**